File: 3sum-closest.py
class Solution:
    # @return an integer
    def threeSumClosest(self, num, target):
        ret = 0
        LL = len(num)
        min_diff = 0
        flag = False
        # sort num
        num.sort()
        # search
        for i in range(LL - 2):
            goal = target - num[i]
            j, k = i + 1, LL - 1
            while j < k:
                tmp = num[j] + num[k]
                sum3 = num[i] + tmp
                if tmp > goal:
                    diff = tmp - goal
                    k -= 1
                elif tmp < goal:
                    diff = goal - tmp
                    j += 1
                else: # tmp == goal
                    return target
                if not flag or diff < min_diff:
                    min_diff = diff
                    ret = sum3
                    flag = True
        return ret

File: 3sum-closest.java
public class Solution {
    public int threeSumClosest(int[] num, int target) {
        int ret = 0;
		int LL = num.length;
		int tmp, goal, diff, sum;
		int minDiff = 0;
		boolean flag = false;
		int i, j, k, g;
		// sort num
		Arrays.sort(num);
		// search
		for (i = 0; i < LL - 2; i++) {
		    goal = target - num[i];
			for (j = i + 1, k = LL - 1; j < k; ) {
				tmp = num[j] + num[k];
				sum = num[i] + tmp;
				if (tmp > goal) {
				    diff = tmp - goal;
				    k--;
				} else if (tmp < goal) {
				    diff = goal - tmp;
				    j++;
				} else { // diff = 0
				    return target;
				}
				if (!flag || diff < minDiff) {
				    minDiff = diff;
				    ret = sum;
				    flag = true;
				}
			}
		}
		return ret;
    }
}